connectors/sources/google_drive.py

```diff
@@ -0,0 +1,626 @@
+        folders = {}
+        async for page in self.get_folders():
+            folders_chunk = page.get("files", [])
+            for folder in folders_chunk:
+                folders[folder["id"]] = {
+                    "name": folder["name"],
+                    "parents": folder.get("parents", None),
+                }
+
+        return folders
+
+    async def resolve_paths(self):
+        """Builds a lookup between a folder id and its absolute path in Google Drive structure
+
+        Returns:
+            dict: mapping between folder id and its (name, parents, path)
+        """
+        folders = await self.retrieve_all_folders()
+        drives = await self.retrieve_all_drives()
+
+        # for paths let's treat drives as top level folders
+        for id, drive_name in drives.items():
+            folders[id] = {"name": drive_name, "parents": []}
+
+        self._logger.info(f"Resolving folder paths for {len(folders)} folders")
+
+        for folder in folders.values():
+            path = [folder["name"]]  # Start with the folder name
+
+            parents = folder["parents"]
+            parent_id = parents[0] if parents else None
+
+            # Traverse the parents until reaching the root or a missing parent
+            while parent_id and parent_id in folders:
+                parent_folder = folders[parent_id]
+                path.insert(
+                    0, parent_folder["name"]
+                )  # Insert parent name at the beginning
+                parents = parent_folder["parents"]
+                parent_id = parents[0] if parents else None
+
+            folder["path"] = "/".join(path)  # Join path elements with '/'
+
+        return folders
+
+    async def _download_content(self, blob, download_func):
+        """Downloads the file from Google Drive and returns the encoded file content.
+
+        Args:
+            blob (dict): Formatted blob document.
+            download_func (partial func): Partial function that gets the file content from Google Drive API.
+
+        Returns:
+            attachment, blob_size (tuple): base64 encoded contnet of the file and size in bytes of the attachment
+        """
+
+        temp_file_name = ""
+        blob_name = blob["name"]
+
+        self._logger.debug(f"Downloading {blob_name}")
+
+        async with NamedTemporaryFile(mode="wb", delete=False) as async_buffer:
+            await download_func(
+                pipe_to=async_buffer,
+            )
+
+            temp_file_name = async_buffer.name
+
+        await asyncio.to_thread(
+            convert_to_b64,
+            source=temp_file_name,
+        )
+
+        blob_size = os.stat(temp_file_name).st_size
+
+        async with aiofiles.open(file=temp_file_name, mode="r") as target_file:
+            attachment = (await target_file.read()).strip()
+
+        self._logger.debug(
+            f"Downloaded {blob_name} with the size of {blob_size} bytes "
+        )
+
+        await remove(str(temp_file_name))
```

Comment:
Personally, I'm not a fan of temp files. In the [official guide](https://developers.google.com/drive/api/guides/manage-downloads#python), there is an example of using IO streaming. Have you considered that approach?

Reply:
Not a blocker. We can revisit later.